src/probnum/diffeq/_odesolver.py

```diff
@@ -1,66 +1,145 @@
 """ODE solver interface."""
 
 from abc import ABC, abstractmethod
+from collections import abc
+from typing import Iterable, Optional, Union
+
+import numpy as np
+
+from probnum import problems
+from probnum.diffeq import callbacks
+from probnum.typing import FloatArgType
+
+CallbackType = Union[callbacks.ODESolverCallback, Iterable[callbacks.ODESolverCallback]]
+"""Callback interface type."""
 
 
 class ODESolver(ABC):
     """Interface for ODE solvers in ProbNum."""
 
-    def __init__(self, order, steprule):
+    def __init__(
+        self,
+        steprule,
+        order,
+    ):
+        self.steprule = steprule
         self.order = order  # e.g.: RK45 has order=5, IBM(q) has order=q
         self.num_steps = 0
-        self.steprule = steprule
 
-    def solve(self, ivp):
-        """Solve an IVP."""
+    def solve(
+        self,
+        ivp: problems.InitialValueProblem,
+        stop_at: Iterable[FloatArgType] = None,
+        callbacks: Optional[CallbackType] = None,
+    ):
+        """Solve an IVP.
+
+        Parameters
+        ----------
+        steprule
+            Step-size selection rule, e.g. constant steps or adaptive steps.
+        stop_at
+            Locations that shall be part of the posterior grid.
+        callbacks
+            Callbacks to happen after every accepted step.
+        """
         times, rvs = [], []
-        for t, rv in self.solution_generator(ivp):
-            times.append(t)
-            rvs.append(rv)
+        for state in self.solution_generator(ivp, stop_at=stop_at, callbacks=callbacks):
+            times.append(state.t)
+            rvs.append(state.rv)
 
         odesol = self.rvlist_to_odesol(times=times, rvs=rvs)
         return self.postprocess(odesol)
 
-    def solution_generator(self, ivp):
+    def solution_generator(
+        self,
+        ivp: problems.InitialValueProblem,
+        stop_at: Iterable[FloatArgType] = None,
+        callbacks: Optional[CallbackType] = None,
+    ):
         """Generate ODE solver steps."""
 
-        t, current_rv = self.initialise(ivp)
+        callbacks, time_stopper = self._process_event_inputs(callbacks, stop_at)
+
+        state = self.initialize(ivp)
+        yield state
+
+        dt = self.steprule.firststep
+
+        # Use state.ivp in case a callback modifies the IVP
+        while state.t < state.ivp.tmax:
+            if time_stopper is not None:
+                dt = time_stopper.adjust_dt_to_time_stamps(state.t, dt)
+
+            state, dt = self.perform_full_step(state, dt)
+
+            if callbacks is not None:
+                for callback in callbacks:
+                    state = callback(state)
+
+            self.num_steps += 1
```

Comment:
Unrelated to the PR, but don't we rather want to count the step attempts for benchmarks? This quantity should also be easy to get, by calling `len` on the solution.

Reply:
Maybe. I am not sure what this quantity is used for at the moment.  The whole "ODE solver statistics" area is rather unexplored at the moment. We can change `self.num_steps` to a variable that counts the attempted steps, but in this case, we should rename the variable. Do you think I should do this here?

Reply:
I think it's fine to leave it out of this PR, it was more of a general comment to get your opinion about this quantity and quantities to track in general :)

Reply:
Alright. Now that I know more about what you were aiming for here: I would like to track many more quantities than that, but when and how this wish is executed is not clear yet. There are more urgent todos :)